runtime/doc/lua-plugin.txt

```diff
@@ -0,0 +1,290 @@
+*lua-plugin.txt*                       Nvim
+
+                            NVIM REFERENCE MANUAL
+
+                    Tutorial on how to write a Lua plugin
+
+
+                                       Type |gO| to see the table of contents.
+
+==============================================================================
+Introduction                                                        *lua-plugin*
+
+Writing a Lua plugin from scratch can be overwhelming. This tutorial is aimed
+to provide basic guidance for it in the form of hands-on example. We will be
+building a plugin called "no-trailing" with scope to provide functionality
+to remove last (i.e. trailing) blank lines (i.e. containing only whitespace)
+in the buffer.
+
+An important disclaimer: Nvim allows flexible approaches to creating Lua
+plugins and is constantly evolving. This tutorial intentionally omits
+providing hard guidelines preferring instead to list some common approaches
+with their pros and cons.
+
+It is a good idea to understand at least basics of the following topics:
+
+• Lua language. See |luaref| and |lua-concepts|.
+• Lua in Nvim. See |lua-guide|.
+• Reading and searching help. See |help.txt| and |:helpgrep|.
+• File system operations, like creating file and directories.
+
+==============================================================================
+Lua code                                                        *lua-plugin-lua*
+
+The core functionality of any Lua plugin is its code. It is located in `lua/`
+subdirectory of the plugin.
+
+First start by creating a directory `no-trailing/` inside your 'runtimepath'.
+In this tutorial let's use the following path:
+`$XDG_DATA_HOME/nvim/site/pack/tutorial/start/no-trailing` (see |$XDG_DATA_HOME
+for its value on your system). The reason to choose this complicated path is
+because it utilizes built-in |packages| functionality to make this plugin
+installed in your config.
+
+The newly created `no-trailing/` directory will be treated as root for all
+relative paths used here. In particular as a first commented line in code
+blocks to describe where it should be placed.
+
+Now let's create a `lua/no-trailing/init.lua` file. It will contain the Lua
+code implementing core functionality. For plugin that removes trailing lines
+it can be something like this:
+>lua
+    -- lua/no-trailing/init.lua
+
+    -- Create a table to store and later return module's user facing functions
+    -- Using `M` is a popular convention, but any name is possible
+    local M = {}
+
+    -- Create user facing function
+    M.trim = function(buf_id, opts)
+      -- Validate required arguments
+      local ok, is_buf_valid = pcall(vim.api.nvim_buf_is_valid, buf_id)
+      assert(ok and is_buf_valid, '`buf_id` is not a valid buffer identifier')
+
+      -- Infer and validate optional arguments
+      opts = vim.tbl_deep_extend('force', { empty = false }, opts or {})
+      assert(type(opts.empty) == 'boolean', '`opts.empty` is not boolean')
```

Comment:
I'd honestly yeet all validation and error checking for the time being, I usually find it annoying when guides go "best practices" on me when I just want the thing to work first.